src/structure.js

```diff
@@ -68,6 +77,57 @@ Structure.prototype.cursor = function (path) {
   return Cursor.from(self.current, path, changeListener);
 };
 
+Structure.prototype.reference = function (path) {
+  var self = this, pathId = pathString(path);
+  var listenerNs = self._pathListeners[pathId];
+
+  var cursor = this.cursor(path);
+  var changeListener = function (newRoot, oldRoot, changedPath) { cursor = self.cursor(path); };
+
+  var referenceListeners = [changeListener];
+  this._pathListeners[path] = !listenerNs ? [changeListener] : listenerNs.concat(changeListener);
+
+  return {
+    onChange: function (newFn) {
+      if (typeof newFn !== 'function') return;
+      var fnIndex, listenerNs = self._pathListeners[pathId];
+
+      self._pathListeners[pathId] = !listenerNs ? [newFn] : listenerNs.concat(newFn);
```

Comment:
if `!listenerNs`, then the invariant should be respected:

> the first element of `self._pathListeners[path]` array is `changeListener` and it is the only element in the array.

thus, should be

`self._pathListeners[pathId] = !listenerNs ? [newFn, changeListener] : listenerNs.concat(newFn);`